test/functional/p2p_blocksonly.py

```diff
@@ -78,13 +62,54 @@ def run_test(self):
             # But if, for some reason, first_peer decides to relay transactions to us anyway, we should relay them to
             # second_peer since we gave relay permission to first_peer.
             # See https://github.com/bitcoin/bitcoin/issues/19943 for details.
-            first_peer.send_message(msg_tx(FromHex(CTransaction(), sigtx)))
+            first_peer.send_message(msg_tx(tx))
             self.log.info('Check that the peer with relay-permission is still connected after sending the transaction')
             assert_equal(first_peer.is_connected, True)
             second_peer.wait_for_tx(txid)
             assert_equal(self.nodes[0].getmempoolinfo()['size'], 1)
         self.log.info("Relay-permission peer's transaction is accepted and relayed")
 
+        self.nodes[0].disconnect_p2ps()
+        self.nodes[0].generate(1)
+
+    def blocks_relay_conn_tests(self):
+        self.log.info('Tests with node in normal mode with block-relay-only connections')
+        self.restart_node(0, ["-noblocksonly"])  # disables blocks only mode
+        assert_equal(self.nodes[0].getnetworkinfo()['localrelay'], True)
+
+        # Ensure we disconnect if a block-relay-only connection sends us a transaction
+        self.nodes[0].add_outbound_p2p_connection(P2PInterface(), connection_type="blockrelay")
+        assert_equal(self.nodes[0].getpeerinfo()[0]['relaytxes'], False)
+        _, txid, tx_hex = self.check_p2p_tx_violation(index=2)
+
+        self.log.info("Check that txs from RPC are not sent to blockrelay connection")
+        conn = self.nodes[0].add_outbound_p2p_connection(P2PTxInvStore(), connection_type="blockrelay")
+
+        # bump time forward to ensure nNextInvSend timer pops
+        self.nodes[0].setmocktime(int(time.time()) + 5)
+
+        assert_equal(self.nodes[0].testmempoolaccept([tx_hex])[0]['allowed'], True)
+        self.nodes[0].sendrawtransaction(tx_hex)
+        conn.sync_with_ping()
+        assert(int(txid, 16) not in conn.get_invs())
```

Comment:
Do we need some kind of sleep/setmocktime here? The node will only send out inventory to outbound connections every 2.5s on average, so this won't catch if we accidentally send out invs to this peer.

Reply:
seems related to your previous feedback https://github.com/bitcoin/bitcoin/pull/19315#discussion_r473860655. thanks for helping me ensure this test is meaningful :) 

the way I've sanity tested this assertion is by removing the `connection_type="blockrelay"` arg to see that it fails. I think the mocktime bump a few lines up should be covering the `INVENTORY_BROADCAST_INTERVAL` delay. is that what you're referring to?

however, even though it appropriately fails (when removing `blockrelay` arg) on my computer, it might still be worth adding a small delay (maybe 0.1 second sleep?) to ensure the `SendMessages` loop is actually hit on every machine? 

what do you think?

Reply:
cc @MarcoFalke, test sync expert 😛

Reply:
wouldn't the setmocktime be called after the tx has been added to the mempool?

Reply:
oh great point. I think its working fine on my machine bc the window between bumping mocktime + making txn is small enough, but I've just pushed an update to have the right order.

Reply:
@MarcoFalke, @jnewbery if you get the chance, lmk if this looks good and I'll resolve. but leaving the thread open for now bc I want a second opinion on whether I've solved the issue

Reply:
The average delay between sending INVs to outbound peers is 2.5s, so bumping mocktime by 5 seconds is only double the mean. Is there any downside in bumping it by a minute so there's overwhelming probability that the node's timer for sending an INV to this peer has popped?

You could also update this to call `sync_with_ping()` twice to ensure that the node has called `SendMessages()` for this peer. Without that, there's a window where the node calls `ProcessMessages()` for the peer, sends the `pong` response and the test moves on to the `assert(int(txid, 16) not in conn.get_invs())` without the node hitting `SendMessages()`. Syncing twice with pings ensures that `ProcessMessages()` is called twice, which means that `SendMessages()` must have been called once.

Reply:
good point & clever trick! I've bumped the mocktime by a minute, added two calls to `sync_with_ping()` and a comment explaining this reasoning. 

thanks for this thoughtful response! going to resolve the comment now, let me know incase there's anything else I should be thinking about.